src/path.rs

```diff
@@ -163,12 +177,26 @@ impl Path {
 
     pub fn from_raw_parts(
         id: usize,
-        points: Vec<PathPoint>,
+        mut points: Vec<PathPoint>,
         trailing: Option<DPoint>,
         closed: bool,
     ) -> Self {
         assert!(!points.is_empty(), "path may not be empty");
         assert!(points.iter().all(|pt| pt.id.parent == id), "{:#?}", points);
+        if !closed {
+            assert!(points.first().unwrap().is_on_curve());
```

Comment:
As a general thought, I think we should export validation errors through the UI rather than asserts. But that can be a matter for another PR.

Reply:
agreed; I'm just feeling a bit sketched out by some of our point-handling logic, and want to be a bit more paranoid.